### ISSUE TITLE
Unhandled exception when stream is cancelled

### ISSUE THREAD

**Issue by a14n** [grpc/grpc-dart#131]
I use `grpc-0.6.5` with the following code:

```dart
main(List<String> args) async {
  await Server([Service()]).serve(port: 8022);

  final client = ServiceClient(ClientChannel(
    'localhost',
    port: 8022,
    options: const ChannelOptions(credentials: ChannelCredentials.insecure()),
  ));

  print(await client.method(Empty()).isEmpty);
  print(await client.method(Empty()).isEmpty); // line 18
}

class Service extends ServiceBase {
  @override
  Stream<Empty> method(ServiceCall call, Empty request) async* {
    yield request;
    yield request;
  }
}
```

and the proto:

```proto
syntax = "proto3";
import "google/protobuf/empty.proto";
service Service {
  rpc Method (google.protobuf.Empty) returns (stream google.protobuf.Empty);
}
```

When I run the code (with `Dart-2.0.0`) I get the following output:

```
false
Unhandled exception:
gRPC Error (14, Error making call: Bad state: The http/2 connection is no longer active and can therefore not be used to make new streams.)
#0      main (file:///home/me/sample/bin/issue-grpc-stream.dart:18:9)
<asynchronous suspension>
#1      _startIsolate.<anonymous closure> (dart:isolate/runtime/libisolate_patch.dart:277:32)
#2      _RawReceivePortImpl._handleMessage (dart:isolate/runtime/libisolate_patch.dart:165:12)
Exited (255)
```

Basically the code starts a server that provides a stream method. I call the method 2 times and use a `.isEmpty` on the resulting stream.

It looks like the problem is about cancelling a stream (because `.isEmpty` cancel its subscription directly after receiving the first event).

(I had a hard time to find what was the issue and I think it's an important issue)

**Comment by a14n** [comment]
/cc @sigurdm 

Maybe related to #111

**Comment by a14n** [comment]
After a hard time to understand how the code behaved it looks like a simple fix is to apply this diff:

```diff
diff --git a/lib/src/client/call.dart b/lib/src/client/call.dart
index cf4dbd8..64c3888 100644
--- a/lib/src/client/call.dart
+++ b/lib/src/client/call.dart
@@ -155,7 +155,10 @@ class ClientCall<Q, R> implements Response {
       }
       _responses.onPause = _responseSubscription.pause;
       _responses.onResume = _responseSubscription.resume;
-      _responses.onCancel = _responseSubscription.cancel;
+      _responses.onCancel = () {
+        cancel();
+        _responseSubscription.cancel();
+      };
     }
   }
```

**Comment by a14n** [comment]
Perhaps even better with:

```diff
diff --git a/lib/src/client/call.dart b/lib/src/client/call.dart
index cf4dbd8..ef1bb26 100644
--- a/lib/src/client/call.dart
+++ b/lib/src/client/call.dart
@@ -155,7 +155,7 @@ class ClientCall<Q, R> implements Response {
       }
       _responses.onPause = _responseSubscription.pause;
       _responses.onResume = _responseSubscription.resume;
-      _responses.onCancel = _responseSubscription.cancel;
+      _responses.onCancel = _safeTerminate;
     }
   }
```

**Comment by ishaanbahal** [comment]
Stumbled on the same issue, couldn't reach server for a good 2-3 minutes thanks to a network provider change on the device. Have you managed to solved it?

**Comment by a14n** [comment]
Not yet unfortunately. The issue seems to be caused by the reception of messages on a closed http2 stream (closed after stream cancellation). This seems to close the connection and prevent subsequent calls. I didn't find a workaround/fix for now. I really hope that a maintainer with better understanding of http2/grpc (like @sigurdm) could take a look because this issue is a show stopper for using this package.

**Comment by a14n** [comment]
I suspect the issue is in [http2](https://github.com/dart-lang/http2) package. IMO frames received in a stream that is in process of closing should be discard. The current version tries to enqueue the message (from DataFrame) to the StreamMessageQueueIn that is _in the process of closing_ and this StreamMessageQueueIn throws an error.

Either the StreamMessageQueueIn should trash the message silently or the message shouldn't reach StreamMessageQueueIn. @sigurdm wdyt?

**Comment by sigurdm** [comment]
That sounds probable. I will try to spend some time on this today.

**Comment by a14n** [comment]
>I will try to spend some time on this today.

Thanks a lot!

**Comment by sigurdm** [comment]
I talked to @mkustermann and he explained that the problem is the grpc package not being able to open a new http2-connection if the current one closes.

He already have written a helper class doing this: https://github.com/dart-lang/appengine/blob/4c9402e254cc921f7951f5c2f1022bd692ccbed6/lib/src/grpc_api_impl/grpc.dart#L298
And here is how it should be used: https://github.com/dart-lang/appengine/blob/4c9402e254cc921f7951f5c2f1022bd692ccbed6/lib/src/grpc_api_impl/grpc.dart#L133
The `connection.isOpen`-check should be before doing the request https://github.com/grpc/grpc-dart/blob/c02c5c46de1cad0a382a9fbf99954dfcb0a06527/lib/src/client/call.dart#L116

I hope to make a patch tomorrow fixing this.

**Comment by a14n** [comment]
>the problem is the grpc package not being able to open a new http2-connection if the current one closes.

It is a problem to fix but IMHO the connection shouldn't have been closed here.

If cancelling a response stream causes the connection's closing it's a problem because other requests could be in progress on the same connection.

**Comment by a14n** [comment]
I made a quick test to confirm what I said. If I cancel a stream while an other stream is sending datas the connection is closed and (worse) the server crash with:

>Bad state: Cannot add event after closing

```dart
// on client side
    client.method(Empty()).forEach((_) => print('-------------${DateTime.now()}'));
    print(await client.method(Empty()).isEmpty); // isEmpty cancels the second stream

// on server side
  Stream<Empty> method(ServiceCall call, Empty request) async* {
    while (true) {
      yield request;
      await Future.delayed(const Duration(seconds: 1));
    }
  }

**Comment by a14n** [comment]
My previous sample was actually working correctly. It's the following code that crash the server.

```dart
// on client side
    client.method(Empty()).forEach((_) => print('-------------${DateTime.now()}'));
    print(await client.method(Empty()).isEmpty); // isEmpty cancels the second stream

// on server side
class Service extends ServiceBase {
  bool first = true;
  @override
  Stream<Empty> method(ServiceCall call, Empty request) async* {
    if (first) {
      first = false;
      while (true) {
        yield request;
        await Future.delayed(const Duration(seconds: 1));
      }
    } else {
      yield request;
      yield request;
      yield request;
    }
  }
}
```

**Comment by sigurdm** [comment]
I think this is the same issue as #111 I really should get that fixed as well?
@mkustermann do you have any insights?

**Comment by mkustermann** [comment]
The https://github.com/grpc/grpc-dart/issues/111 issue seems be due to the `package:grpc` not correctly respecting the `onCancel` event propagated to `http2Stream.outgoingMessages`: If the client cancels/terminates the stream, it's no longer interested in messages, this signal is forwarded to the server side via a cancel on the `outgoingMessages` StreamSink. Adding to a cancelled sink is an error.

**Comment by a14n** [comment]
>I think this is the same issue as #111

Not exactly because in my case I have 2 streams and it's not the infinite stream that is cancelled.

**Comment by a14n** [comment]
@mkustermann the problem seems also in http2 package.

With the following events:

- client call a method that streams results
- server starts the stream an send the first DataFrame
- client receives the first DataFrame 
- client cancel the stream
- server (that have not yet received the cancellation) send the second DataFrame
- server received cancellation and stop sending DataFrame
- client receives the second DataFrame

The last line leads today to an error (grabbed with some additionnal logs):

```
Connection._catchProtocolErrors: Bad state: Was in the process of closing.
#0      _StreamMessageQueueIn&Object&TerminatableMixin&ClosableMixin.ensureNotClosingSync (package:http2/src/error_handler.dart:73:7)
#1      StreamMessageQueueIn.enqueueMessage (package:http2/src/flowcontrol/stream_queues.dart:217:5)
#2      ConnectionMessageQueueIn._tryDispatch (package:http2/src/flowcontrol/connection_queues.dart:317:10)
#3      ConnectionMessageQueueIn._addMessage (package:http2/src/flowcontrol/connection_queues.dart:295:5)
#4      ConnectionMessageQueueIn.processDataFrame (package:http2/src/flowcontrol/connection_queues.dart:253:5)
#5      StreamHandler._handleDataFrame (package:http2/src/streams/stream_handler.dart:634:19)
#6      StreamHandler._processStreamFrameInternal.<anonymous closure> (package:http2/src/streams/stream_handler.dart:596:11)
#7      _StreamHandler&Object&TerminatableMixin.ensureNotTerminatedSync (package:http2/src/error_handler.dart:33:13)
#8      StreamHandler._processStreamFrameInternal (package:http2/src/streams/stream_handler.dart:509:12)
#9      StreamHandler.processStreamFrame (package:http2/src/streams/stream_handler.dart:469:7)
#10     Connection._handleFrameImpl (package:http2/src/connection.dart:368:16)
#11     Connection._setupConnection.<anonymous closure>.<anonymous closure> (package:http2/src/connection.dart:155:34)
#12     Connection._catchProtocolErrors (package:http2/src/connection.dart:300:9)
#13     Connection._setupConnection.<anonymous closure> (package:http2/src/connection.dart:155:7)
#14     StackZoneSpecification._registerUnaryCallback.<anonymous closure>.<anonymous closure> (package:stack_trace/src/stack_zone_specification.dart:129:26)
#15     StackZoneSpecification._run (package:stack_trace/src/stack_zone_specification.dart:209:15)
#16     StackZoneSpecification._registerUnaryCallback.<anonymous closure> (package:stack_trace/src/stack_zone_specification.dart:129:14)
#17     _rootRunUnary (dart:async/zone.dart:1132:38)
#18     _CustomZone.runUnary (dart:async/zone.dart:1029:19)
#19     _CustomZone.runUnaryGuarded (dart:async/zone.dart:931:7)
#20     _BufferingStreamSubscription._sendData (dart:async/stream_impl.dart:336:11)
#21     _DelayedData.perform (dart:async/stream_impl.dart:584:14)
#22     _StreamImplEvents.handleNext (dart:async/stream_impl.dart:700:11)
#23     _PendingEvents.schedule.<anonymous closure> (dart:async/stream_impl.dart:660:7)
#24     StackZoneSpecification._run (package:stack_trace/src/stack_zone_specification.dart:209:15)
#25     StackZoneSpecification._registerCallback.<anonymous closure> (package:stack_trace/src/stack_zone_specification.dart:119:48)
#26     _rootRun (dart:async/zone.dart:1120:38)
#27     _CustomZone.run (dart:async/zone.dart:1021:19)
#28     _CustomZone.runGuarded (dart:async/zone.dart:923:7)
#29     _CustomZone.bindCallbackGuarded.<anonymous closure> (dart:async/zone.dart:963:23)
#30     StackZoneSpecification._run (package:stack_trace/src/stack_zone_specification.dart:209:15)
#31     StackZoneSpecification._registerCallback.<anonymous closure> (package:stack_trace/src/stack_zone_specification.dart:119:48)
#32     _rootRun (dart:async/zone.dart:1124:13)
#33     _CustomZone.run (dart:async/zone.dart:1021:19)
#34     _CustomZone.runGuarded (dart:async/zone.dart:923:7)
#35     _CustomZone.bindCallbackGuarded.<anonymous closure> (dart:async/zone.dart:963:23)
#36     _microtaskLoop (dart:async/schedule_microtask.dart:41:21)
#37     _startMicrotaskLoop (dart:async/schedule_microtask.dart:50:5)
#38     _runPendingImmediateCallback (dart:isolate/runtime/libisolate_patch.dart:113:13)
#39     _RawReceivePortImpl._handleMessage (dart:isolate/runtime/libisolate_patch.dart:166:5)
```

This leads to connection (and all its streams) closing.

IMO this second DataFrame shoud be discard. In the [http2 spec](https://http2.github.io/http2-spec/#RST_STREAM) I see:

>However, after sending the RST_STREAM, the sending endpoint MUST be prepared to receive and process additional frames sent on the stream that might have been sent by the peer prior to the arrival of the RST_STREAM. 

WDYT?

**Comment by mkustermann** [comment]
@a14n Yes, that particular scenario seems like a bug in `package:http2`. If you happen to have a reproduction, could you file a new issue with it in https://github.com/dart-lang/http2/issues ?

**Comment by a14n** [comment]
Unfortunately I failed at create a simple reproduction ( async calls are horrible to follow :-( ).

For now it looks like the cancellation of `streamQueueIn` is not tracked and is handled the same way as if the stream is closed.

I made [changes at several places](https://github.com/dart-lang/http2/compare/master...a14n:accept-incoming-frame-after-cancel) but they are not enough to fix the problem.

There are still issue with the removal of the stream from `StreamHandler._openStreams` by https://github.com/dart-lang/http2/blob/34fc7780a196be782f50c20ccbc333075c7cb8f0/lib/src/streams/stream_handler.dart#L336-L338 . Once called the next incoming frame [doesn't match any open stream](https://github.com/dart-lang/http2/blob/34fc7780a196be782f50c20ccbc333075c7cb8f0/lib/src/streams/stream_handler.dart#L511) and this leads to a protocol error. It looks like cancelled streams should also be tracked in this class.

I don't know enough the http2 code to make a good PR and it would be really awesome if you could take a look.

**Comment by mkustermann** [comment]
I've filed https://github.com/dart-lang/http2/issues/34 for the http2 issue.

**Comment by a14n** [comment]
Fixed by dart-lang/http2#35 available in package _http2-0.1.9_.
